src/Interpreters/Cache/QueryResultCache.cpp

```diff
@@ -0,0 +1,343 @@
+/// more natural if settings related to the query result cache are erased from the AST key. Note that at this point the settings themselves
+/// have been parsed already, they are not lost or discarded.
+ASTPtr removeQueryResultCacheSettings(ASTPtr ast)
+{
+    ASTPtr transformed_ast = ast->clone();
+
+    RemoveQueryResultCacheSettingsMatcher::Data visitor_data;
+    RemoveQueryResultCacheSettingsVisitor(visitor_data).visit(transformed_ast);
+
+    return transformed_ast;
+}
+
+}
+
+QueryResultCache::Key::Key(
+    ASTPtr ast_, String partition_key_,
+    Block header_, const std::optional<String> & username_,
+    std::chrono::time_point<std::chrono::system_clock> expires_at_)
+    : ast(removeQueryResultCacheSettings(ast_))
+    , partition_key(partition_key_)
+    , header(header_)
+    , username(username_)
+    , expires_at(expires_at_)
+{
+}
+
+bool QueryResultCache::Key::operator==(const Key & other) const
+{
+    return ast->getTreeHash() == other.ast->getTreeHash()
+        && partition_key == other.partition_key;
+}
+
+String QueryResultCache::Key::queryStringFromAst() const
+{
+    WriteBufferFromOwnString buf;
+    IAST::FormatSettings format_settings(buf, /*one_line*/ true);
+    ast->format(format_settings);
+    return buf.str();
+}
+
+size_t QueryResultCache::KeyHasher::operator()(const Key & key) const
+{
+    SipHash hash;
+    hash.update(key.ast->getTreeHash());
+    hash.update(key.partition_key);
+    auto res = hash.get64();
+    return res;
+}
+
+namespace
+{
+
+auto is_stale = [](const QueryResultCache::Key & key)
+{
+    return (key.expires_at < std::chrono::system_clock::now());
+};
+
+}
+
+QueryResultCache::Writer::Writer(std::mutex & mutex_, Cache & cache_, const Key & key_,
+    size_t & cache_size_in_bytes_, size_t max_cache_size_in_bytes_,
+    size_t max_cache_entries_,
+    size_t max_entry_size_in_bytes_, size_t max_entry_size_in_rows_,
+    std::chrono::milliseconds min_query_runtime_)
+    : mutex(mutex_)
+    , cache(cache_)
+    , key(key_)
+    , cache_size_in_bytes(cache_size_in_bytes_)
+    , max_cache_size_in_bytes(max_cache_size_in_bytes_)
+    , max_cache_entries(max_cache_entries_)
+    , max_entry_size_in_bytes(max_entry_size_in_bytes_)
+    , max_entry_size_in_rows(max_entry_size_in_rows_)
+    , min_query_runtime(min_query_runtime_)
+{
+    if (auto it = cache.find(key); it != cache.end() && !is_stale(it->first))
+        skip_insert = true; /// Key already contained in cache and did not expire yet --> don't replace it
+}
+
+void QueryResultCache::Writer::finalizeWrite()
+{
+    if (skip_insert)
+        return;
+
+    if (std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now() - query_start_time) < min_query_runtime)
+        return;
+
+    auto to_single_chunk = [](const Chunks & chunks_) -> Chunk
+    {
+        if (chunks_.empty())
+            return {};
+
+        Chunk res = chunks_[0].clone();
```

Comment:
`chunks_[0]` could be moved from if I'm not mistaken

Reply:
You are right ;)